tracing-attributes/src/lib.rs

```diff
@@ -579,3 +663,80 @@ fn instrument_err(args: &[NestedMeta]) -> bool {
         _ => false,
     })
 }
+
+// terrible machinery to make async_trait-like cases works
+// (following the approach suggested in
+// https://github.com/dtolnay/async-trait/issues/45#issuecomment-571245673)
+fn is_async_trait(block: &Block, block_is_async: bool) -> Option<String> {
+    // are we in an async context ? If yes, this isn't a async_trait-like pattern
+    if block_is_async {
+        return None;
+    }
+
+    // list of async functions declared inside the block
+    let mut inside_funs = Vec::new();
+    // last expression declared in the block (it determines the return value of the block, so
+    // that if we are working on a function whose `trait` or `impl` declaration is annotated by
+    // async_trait, this is quite likely the point where the future is pinned)
+    let mut last_expr = None;
+
+    // obtain the list of direct internal functions and the last expression of the block
+    for stmt in &block.stmts {
+        if let Stmt::Item(Item::Fn(fun)) = &stmt {
+            // is the function declared as async ? If so, this is a good candidate, let's keep it
+            // in hand
+            if fun.sig.asyncness.is_some() {
+                inside_funs.push(fun.sig.ident.to_string());
+            }
+        } else if let Stmt::Expr(e) = &stmt {
+            last_expr = Some(e);
+        }
+    }
+
+    // let's play with (too much) pattern matching
+    // is the last expression a function call ?
+    if let Some(Expr::Call(ExprCall {
+        func: outside_func,
+        args: outside_args,
+        ..
+    })) = last_expr
+    {
+        if let Expr::Path(path) = outside_func.as_ref() {
+            // is it a call to `Box::pin()` ?
+            if "Box::pin"
+                == path
+                    .path
+                    .segments
+                    .iter()
+                    .map(|x| x.ident.to_string())
+                    .collect::<Vec<String>>()
```

Comment:
I think we could avoid a couple of allocations if we did this by creating a string and used `push_str` to append the string slices from the `path.segments` iterator, rather than creating a `String` from each string slice and collecting them into a `Vec`. I don't _think_ this is hot enough that this is terribly important though, so it's fine as-is.